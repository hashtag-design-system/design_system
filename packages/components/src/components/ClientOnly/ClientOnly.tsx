import React from "react";
import { useHasMounted } from "../../utils";

// Big thanks to -> https://www.joshwcomeau.com/react/the-perils-of-rehydration/

export type Props = {
  defaultMounted?: boolean;
};

const ClientOnly: React.FC<Props> = ({ defaultMounted = false, children }) => {
  // const [hasMounted, setHasMounted] = useState(defaultMounted);
  const [hasMounted] = useHasMounted(defaultMounted);

  if (!hasMounted) return null;

  return <>{children}</>;
};

ClientOnly.displayName = "ClientOnly";

export default ClientOnly;
