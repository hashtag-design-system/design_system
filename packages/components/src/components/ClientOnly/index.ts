import ClientOnly from "./ClientOnly";

export { default as ClientOnly } from "./ClientOnly";
export type { Props as ClientOnlyProps } from "./ClientOnly";

export default ClientOnly;