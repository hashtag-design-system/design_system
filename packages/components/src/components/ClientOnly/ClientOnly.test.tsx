import { render, screen } from "@testing-library/react";
import ClientOnly from "./index";

describe("<ClientOnly />", () => {
  test.each([undefined, true])("default behaviour & with defaultMounted", defaultMounted => {
    render(
      <ClientOnly defaultMounted={defaultMounted}>
        <div data-testid="div">Hey!</div>
      </ClientOnly>,
      { hydrate: !defaultMounted ? true : false  }
    );

    if (typeof window !== "undefined") {
      const div = screen.getByTestId("div");
      
      expect(div).toBeVisible();
      expect(div).toHaveTextContent("Hey!");
      expect(div.children).toHaveLength(0);
      expect(div).toMatchSnapshot();
    }
  });
});
