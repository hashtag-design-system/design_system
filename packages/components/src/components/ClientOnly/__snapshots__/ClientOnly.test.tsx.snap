// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientOnly /> default behaviour & with defaultMounted 1`] = `
<div
  data-testid="div"
>
  Hey!
</div>
`;

exports[`<ClientOnly /> default behaviour & with defaultMounted 2`] = `
<div
  data-testid="div"
>
  Hey!
</div>
`;
